### ISSUE TITLE
Game project becomes inconsistent if not saved after deleting translation

### ISSUE THREAD

**Issue by Crystal-Shard** [adventuregamestudio/ags#2414]
**Describe the bug**
Sometimes the editor will save the Game.agf file when the user doesn't ask to. This behavior is inconsistent and can break the game.

**AGS Version**
3.6.0

**Game**
n/a

**To Reproduce**
1. Open any game
2. Create a new translation
3. save now (I had missed the fact that filling the TRS file automatically saves the game)
4. Delete the translation
5. Quit the editor without saving
6. It turns out that the game was quietly saved after step 2 anyway; following these steps will result in an unreadable game.agf file

**Expected behavior**
Either (1) do not save unless the user presses save, or (2) be consistent when to save (e.g. BOTH when creating a translation AND when deleting the translation).

**Desktop (please complete the following information):**
 - OS: Windows
 - Version 11

**Comment by ericoporto** [comment]
> following these steps will result in an unreadable game.agf file

I think this was fixed in AGS 3.6.1 already.

**Comment by ivan-mogilko** [comment]
Editor loading with missing TRS was fixed in 3.6.1.

But I was not aware that it saves the project after adding translation. This should be fixed, IMO. It does not save when a room is added, instead it warns when closing the editor without saving, saying something like "files have been added to project, if you dont save now, your game wont remember them" etc. Similar thing could be done for translations. Perhaps, fonts too, since we are at that.

**Comment by ivan-mogilko** [comment]
I tried to replicate this in 3.6.0.

Editor does NOT save when adding translations. In fact, it warns you on exit if translation was created, but game not saved:

>
> Save changes?
>
> Files have been added, removed or renamed. If you don't save the game now, you may not be able to successfully open this game next time. Do you want to save your changes?

Same happens if you create and then instantly delete a translation, then quit.

Editor saves when you do "Update Translation", and warns about that too:

> Updating the translation can take some time depending on the size of your game, and **will save your game beforehand**. Do you want to continue?

**Comment by Crystal-Shard** [comment]
Ok, I've double-checked and the editor does give warnings. But I do find this approach problematic;

First, when creating a translation, I get a query whether I'd like to fill the new translation with existing text. Note that this is a false choice for the user: if I _don't_ do this, then the translation is completely pointless.

Second, when exiting the editor, I get a warning that if I don't save now, this might break the game. Note that again this is a false choice: if I _don't_ do this, then I end up with a broken and unusable .AGF file.

I'm not a fan of choices like these. If there's and A-or-B choice and choosing A will cause the game (or translation) to fail, then this should not be asked of the user, and B should be automatic.

**Comment by ericoporto** [comment]
Can you create a small example of such broken agf file? I can't figure out any case that causes this with the provided steps.

**Comment by ivan-mogilko** [comment]
> Second, when exiting the editor, I get a warning that if I don't save now, this might break the game. Note that again this is a false choice: if I don't do this, then I end up with a broken and unusable .AGF file.

This is not accurate, if you don't save, then agf won't be broken, there will be a unused translation file remaining in a folder, but agf won't contain anything invalid.

**Comment by Crystal-Shard** [comment]
> This is not accurate, if you don't save, then agf won't be broken, there will be a unused translation file remaining in a folder, but agf won't contain anything invalid.

It's the reverse: if you delete the TRS from the editor, then it gets removed from disk. If you then do not save the game, then the .AGF file will contain a reference to a file that does not exist, and will refuse to load (although as mentioned above, in 3.6.1 this is fixed).

**Comment by ivan-mogilko** [comment]
Okay, I see now, so this is when you
- created Translation
- saved the game
- deleted Translation
- quit without saving.

Probably same may happen with other items, like rooms and fonts.

**Comment by Crystal-Shard** [comment]
> Okay, I see now, so this is when you
> 
> * created Translation
> * saved the game
> * deleted Translation
> * quit without saving.

Yes. So my issue here is with the warning "If you don't save the game now, you may not be able to successfully open this game next time."

Instead of this warning, simply save the game. The game is also saved automatically when you compile, when you fill a TRS file, and probably in a few other situations. It's not really helpful to ask the user "If you don't do X, your game file will be unusable. Do you want to X?"

**Comment by ivan-mogilko** [comment]
Alright, so there are two things
1. Make sure that Editor does not ask wrong questions.
2. Make sure that the project may be loaded if some optional parts are missing. This is a standalone task, because these files may get deleted by hand.

Loading a project with missing translations was fixed in 3.6.1.
I recall that I also fixed loading with missing scripts, either in 3.6.0 or 3.6.1.
What is left to check:
- fonts;
- rooms.

Regarding the version, since 3.6.1 was released, I'm now not patching previous versions like 3.6.0, unless there's a critical error that cannot be worked around.

**Comment by ericoporto** [comment]
@Crystal-Shard can you 100% reproduce this in the latest 3.6.1 Editor?

**Comment by Crystal-Shard** [comment]
> Regarding the version, since 3.6.1 was released, I'm now not patching previous versions like 3.6.0, unless there's a critical error that cannot be worked around.

That's fine. I'm simply reporting bugs for the AGS version I'm currently using, I don't always keep pace with the latest builds. If an issue is important to me and has been fixed in a later engine, well that's a good reason for me to upgrade :)

**Comment by ivan-mogilko** [comment]
> can you 100% reproduce this in the latest 3.6.1 Editor?

The problem reproduces in 3.6.1 using above steps, with the exception of editor failing to load. Game.agf keeps mentioning a non-existing translation file, and it gets recreated after loading a project.

**Comment by ivan-mogilko** [comment]
> Yes. So my issue here is with the warning "If you don't save the game now, you may not be able to successfully open this game next time."

Instead of this warning, simply save the game. The game is also saved automatically when you compile, when you fill a TRS file, and probably in a few other situations. It's not really helpful to ask the user "If you don't do X, your game file will be unusable. Do you want to X?"


In regards to saving the project unconditionally, I would not take such step lightly. There are 2 considerations here:
1. User may have done other changes to the project prior to adding or deleting translation, and does not want to save these changes. If we just save unconditionally, that will go against user's expectations and may damage their work. This is a rare case, but I see how it's possible.
2. There are few actions that create new files in the project now, but I may imagine a hypothetical situation where adding anything to project corresponds to adding a file. This is a matter of asset management and storage, and may change from version to version. If we take as a principle, that whenever something adds a file, a game implicitly saves, then such principle may cause problems in the future.

For this reason, I'd rather suggest going other way, not save without a question, but ensure that Editor:
1. Can clean the temporary project files up when not saving.
2. Can restore project's state on load if anything is missing.

**Comment by Crystal-Shard** [comment]
Well it's true that saving unconditionally shouldn't be done lightly... but the editor _already_ saves unconditionally whenever you build the game, or rebuild one of the translation sources, or probably in some other cases. So maybe it's not such a big deal to save unconditionally?

**Comment by ivan-mogilko** [comment]
I don't know what is the right way to deal with this just yet. I may be overthinking this case, but my head is occupied with other things now; and I'd like to discuss this with other developers first. 

What should be certainly done in the meantime, is double check that Editor can recover the project if any auxiliary files are missing, by either reimporting existing ones or creating a dummy empty replacement.

**Comment by ivan-mogilko** [comment]
I gave this some more thought, and in the end remain at the point explained in one of the previous comments:
https://github.com/adventuregamestudio/ags/issues/2414#issuecomment-2102530418

To reiterate, there are several cases already in the editor where it may create and delete files, not only translations, but also rooms, and scripts, and fonts, and maybe will do more of these in the later versions. If we only save on deleting a translation, then this behavior will be inconsistent with the rest. If we save unconditionally whenever user deletes an item synced with a file (so in effect deletes a file too), then we take a control over the project from the user, preventing them from cancelling latest changes would they want to.

I should underline again, that currently editor never saves the game without warning a user first. That at least lets user to rethink what they are doing, if they are unsure.

Furthermore, in the future we might introduce a proper undo feature to the editor, where the deleted files will not be deleted, but, for example, moved to a temporary location, and user will have an opportunity to revert their deletion while in the same editor session.

For the time being, in 3.6.1/3.6.2 Editors already support loading a project with missing files, and recreate blank files for items which need them (such as scripts). So the above situation will not lead to a project corruption or major problems anymore.

I will close this issue as not planned.

**Comment by ivan-mogilko** [comment]
I gave this some more thought, and *in regards to an unconditional saving* I remain at the point explained in one of the previous comments:
https://github.com/adventuregamestudio/ags/issues/2414#issuecomment-2102530418

I don't want to save the game when user does not express a clear intent or agreement to save it. There are several cases already in the editor where it may create and delete files, not only translations, but also rooms, and scripts, and fonts, and maybe will do more of these in the later versions. If we only save on deleting a translation, then this behavior will be inconsistent with the rest. If we save unconditionally whenever user deletes an item synced with a file (so in effect deletes a file too), then we take a control over the project from the user, preventing them from cancelling latest changes would they want to.

I should underline again, that currently editor never saves the game without letting a user know that it's going to do so. That at least lets user to rethink what they are doing, if they are unsure.

---

That question aside, there are following issues remaining, in the order of severity:

**Issue 1:** Having deleted a file (a project item, synced with a file), but not saving the project, will make project *unreadable* next time it is open.
This issue has already *been fixed in AGS 3.6.1*, where it will load the project even if some files are missing, and recreate blank files would they be needed for some project components (translations, scripts, etc).
One thing that was not fixed is the warning that is given when closing without saving:
> Files have been added, removed or renamed. If you don't save the game now, you may not be able to successfully open this game next time.
That's not accurate now, so the text should be adjusted.

**Issue 2:** Having deleted a project item, synced with a file, but not saving the project, will make project *inconsistent*. In the sense that the item will still exist, but file will not. This means that only part of user's action was reverted by not saving.
In order to fix this, we will have to introduce a proper undo feature to the editor, where the deleted files will not be deleted, but, for example, moved to a temporary location, and user will have an opportunity to revert their deletion while in the same editor session.
I might open a feature suggestion for this.

**Issue 3:** Updating the translation requires to save the project. I don't know if that alone is a problem though. It does warn about having to save, and lets user reconsider. The game has to be saved, because Editor needs to gather texts from various places, such as game properties, rooms, scripts, etc. If we do not save, then we'll have to scan the objects in runtime memory for these. But we have future plan to move various processing to standalone tools that are run as separate processes (either from the editor, or from command line) and work strictly with files (input and output). Would translation update be done by these, we will have to save the project files anyway.

**Issue 4:** Editor asking user questions that seem redundant. This was mentioned in the discussion above.
One instance is the question about whether user wants to update translation right after creating a new one. I suppose that given that may be a lengthy process and that requires to save a game, then it does make sense to warn a user.
Second instance, is the question on exit where editor warns that "files were added or deleted" and if user won't save, then project may not load. I agree that this is a strange question, but fixing this question won't solve a real problem. The problem is combined Issue 1 and Issue 2. Issue 1 is already fixed, the project will be able to load even if user did not save. The Issue 2 suggests that we should have a proper undo mechanic, which would revert any added/deleted files would user not save the project.

So, in the end, I think I shall close this ticket as partially solved (by letting to open a project even with missing files),
and open a new ticket with proposal to implement a undo mechanic for project files.

**Comment by AlanDrake** [comment]
Things like adding/removing elements to the project (room/gui/scripts/etc) should proably be immediately saved without asking.
Godot and Visual Studio are the same, the only things that don't get saved immediately are scripts/objects that you can open a tab and undo before committing.

**Comment by ivan-mogilko** [comment]
> Things like adding/removing elements to the project (room/gui/scripts/etc) should proably be immediately saved without asking.
Godot and Visual Studio are the same

Idk about Godot, but Visual Studio does not save the project when adding or removing files.

I have a different opinion on this, which I explained above. Besides being unable to revert, my concern is that if, theoretically, at some point, every global project item becomes a separate file, then if we have this rule to save, it will in effect force editor to save whenever anything is added or removed. In other words, such rule will make project saving depend on how the items are represented (part of a bigger file or a separate file).

I wanted to write a ticket, suggesting a undo mechanic, which would revert adding/removing files along with the project changes.

But if saving project unconditionally is what users or other developers would prefer, then this has to be done consistently and set as a rule for the IDE. This may be discussed on forums too.

**Comment by ivan-mogilko** [comment]
But then, I also think that having a proper Undo/Redo functionality in the project may be of a greater priority, and would cover many use cases. For instance, right now if you have deleted a project item by mistake, then the only way to recover it is to quit without saving and reload a project. If user could just do Ctrl+Z to get it back, that would be more helpful.

**Comment by AlanDrake** [comment]
But how would that work on project level? What if I create a module, put stuff inside, then Undo a bit too much for something else, and find my module gone later? Could I undo by mistake or there would be a specific button on the UX for specifically project-level undo/redo?

I'm not saying it shouldnt be done, but a undo/redo outside the context of individual tabs is not a trivial thing. The tree-like nature of the project may hide the feedback of the action.

**Comment by AlanDrake** [comment]
Perhaps a History panel like Photoshop? I can't think of any other implementation that gives enough information to the user.
It's basically a chronological list of states, with a description of the operation that caused them.
